anndata/experimental/merge.py

```diff
@@ -0,0 +1,615 @@
+        )
+
+
+def _write_concat_sequence(
+    arrays: Sequence[Union[pd.DataFrame, SparseDataset, H5Array, ZarrArray]],
+    output_group,
+    output_path,
+    max_loaded_sparse_elems,
+    axis=0,
+    index=None,
+    reindexers=None,
+    fill_value=None,
+    join="inner",
+):
+    """
+    array, dataframe, csc_matrix, csc_matrix
+    """
+    if any(isinstance(a, pd.DataFrame) for a in arrays):
+        if reindexers is None:
+            if join == "inner":
+                reindexers = gen_inner_reindexers(arrays, None, axis=axis)
+            else:
+                raise NotImplementedError("Cannot reindex dataframes with outer join.")
+        if not all(
+            isinstance(a, pd.DataFrame) or a is MissingVal or 0 in a.shape
+            for a in arrays
+        ):
+            raise NotImplementedError(
+                "Cannot concatenate a dataframe with other array types."
+            )
+        df = concat_arrays(
+            arrays=arrays,
+            reindexers=reindexers,
+            axis=axis,
+            index=index,
+            fill_value=fill_value,
+        )
+        write_elem(output_group, output_path, df)
+    elif all(
+        isinstance(a, (pd.DataFrame, SparseDataset, H5Array, ZarrArray)) for a in arrays
+    ):
+        _write_concat_arrays(
+            arrays,
+            output_group,
+            output_path,
+            max_loaded_sparse_elems,
+            axis,
+            reindexers,
+            fill_value,
+            join,
+        )
+    else:
+        raise NotImplementedError(
+            f"Concatenation of these types is not yet implemented: {[type(a) for a in arrays] } with axis={axis}."
+        )
+
+
+def _write_alt_mapping(groups, output_group, alt_dim, alt_indices, merge):
+    alt_mapping = merge([read_as_backed(g[alt_dim]) for g in groups])
+    # If its empty, we need to write an empty dataframe with the correct index
+    if not alt_mapping:
+        alt_df = pd.DataFrame(index=alt_indices)
+        write_elem(output_group, alt_dim, alt_df)
+    else:
+        write_elem(output_group, alt_dim, alt_mapping)
+
+
+def _write_alt_annot(groups, output_group, alt_dim, alt_indices, merge):
+    # Annotation for other axis
+    alt_annot = merge_dataframes(
+        [read_elem(g[alt_dim]) for g in groups], alt_indices, merge
+    )
+    write_elem(output_group, alt_dim, alt_annot)
+
+
+def _write_dim_annot(groups, output_group, dim, concat_indices, label, label_col, join):
+    concat_annot = pd.concat(
+        unify_dtypes([read_elem(g[dim]) for g in groups]),
+        join=join,
+        ignore_index=True,
+    )
+    concat_annot.index = concat_indices
+    if label is not None:
+        concat_annot[label] = label_col
+    write_elem(output_group, dim, concat_annot)
+
+
+def concat_on_disk(
+    in_files: Union[
+        Collection[Union[str, os.PathLike]],
+        typing.MutableMapping,
+    ],
+    out_file: Union[str, os.PathLike],
+    overwrite: bool = False,
```

Comment:
bool arguments should only be positional when they’re the only poitional argument, so the `*` should be moved before this parameter.

maybe the name `max_loaded_sparse_elems` should be shortened so it can be more easily specified by name too then.